### PR TITLE
PLAT-1211 Transaction Data Deleter Mechanism

### PR DESCRIPTION
**Pull request by oliver-richers**
deletion-of-transactional-data
only available for test systems:
image
not available for non-test systems:
image

slightly changed popup:
image

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java [hunk 1]
```diff
@@ -348,6 +352,8 @@ public interface CoreI18n extends EmfI18n {
 		.de("Externe Komponenten");
 	I18n0 FAILED = new I18n0("Failed")//
 		.de("Fehlgeschlagen");
+	I18n1 FAILED_TO_DELETE_DATA_FROM_TABLE_ARG1 = new I18n1("Failed to delete data from table '%s'.")//
+		.de("Daten aus der Tabelle '%s' konnten nicht gelöscht werden");
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:356]:
```suggestion
		.de("Daten aus der Tabelle '%s' konnten nicht gelöscht werden.");
```

File: platform-core-module/src/main/java/com/softicar/platform/core/module/module/IModule.java [hunk 2]
```diff
@@ -17,4 +19,6 @@
 	 * @return the module instance table (never <i>null</i>)
 	 */
 	IModuleInstanceTable<I> getModuleInstanceTable();
+
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/module/IModule.java:22]: This should have Javadoc

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transactional/data/TransactionalDataDeleter.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import com.softicar.platform.db.runtime.table.dependency.DbTableDependencyGraph;
+import com.softicar.platform.emf.module.registry.CurrentEmfModuleRegistry;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class TransactionalDataDeleter {
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/transactional/data/TransactionalDataDeleter.java:17]: As this class poses a significant threat to the DB content (by design), we should have tests to ensure that it deletes _exactly_ what it's supposed to.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/module/IModule.java (hunk 2 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/module/IModule.java:22]: added Javadoc

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transactional/data/TransactionalDataDeleter.java (hunk 3 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/transactional/data/TransactionalDataDeleter.java]: - hmm, can hardly be tested without payload modules...any suggestions?
- changed popup to display the statements that will be executed, IMO, it's the obligation of the admin to verify those
image

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeleter.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class TransactionDataDeleter {
+
+	public void execute() {
+
+		for (IDbTable<?, ?> table: getTableList()) {
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeleter.java:20]: We could extract the content of this loop to a private method, to improve legibility.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeletionPopup.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		textArea.setSize(25, 80);
+		return textArea;
+	}
+
+	private String getStatements() {
+
+		return deleter//
+			.getTableList()
+			.stream()
+			.map(table -> "DELETE FROM %s;".formatted(table.getFullName()))
```
Comment (alexander-schmidt) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeletionPopup.java:52]: We could let `TransactionDataDeleter` provide those statements.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeleter.java (hunk 4 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeleter.java]: I do not see much value in it, but did so

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeletionPopup.java (hunk 5 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/transaction/data/TransactionDataDeletionPopup.java]: done

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/transactional/data/TransactionalDataDeleter.java (hunk 3 above)
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/transactional/data/TransactionalDataDeleter.java]: implemented test in code module

**Review by alexander-schmidt — APPROVED** [review]

**Pull request merged**